standards/HEALTH_CHECK_SPECIFICATION.md

```diff
@@ -0,0 +1,101 @@
+* The health check must return [status information as JSON](#health-check-body)
+
+### Health check status codes
+
+There are three standard status codes that a health check can return:
+
+Status code | Description
+------------|------------------------
+`200`       | I'm fine
+`429`       | I'm warming up or I'm degraded
+`500`       | Kill me
+
+The health check should return 429 until the first checks have been run.  After that, the codes returned should be based on the check status.  It is the responsibility of whatever is monitoring the health check to provide a grace period on startup.
+
+If any check returns a `CRITICAL` status (see [check statuses below](#check-statuses) then the app should transition to a `429`.  After a timeout that can be set per app, the `429` will be downgraded `500`.  This delay in marking the app as `500` is to allow the issue a chance to resolve on its own before moving to a `500` which will result in the app being killed.
+
+If any check returns a `WARNING` status (see [check statuses below](#check-statuses) then the app should transition to a `429`.  In this situation, the timeout does not apply and the app will continue to return a `429` until the status of the checks changes.
+
+Any status code ≥ 400 && != 429 will be treated as a failure.
+
+### Health check body
+
+The JSON body of the health check should implement the following spec:
+
+Field        | Type     | Description
+-------------|----------|---------------
+`status`     | `string` | The overall status of the health check using the values as the [check statuses](#check-statuses)
+`version`    | `string` | The version information of the app (e.g. commit, semver version, go version, build time, etc) as a string.
+`uptime`     | `ms`     | Milliseconds elapsed since the app started <sup>1</sup>
+`start_time` | `ISO8601`<sup>2</sup> | The time the app started in UTC
+`checks`     | [`[]Check`](#check)   | An array of the checks with details of their statuses
+
+<sup>1</sup> The start time of the app is approximate as we use the time at which the health check library was instantiated.
+
+<sup>2</sup> `ISO8601` UTC date time (`2006-01-02T15:04:05.999Z`)
+
+**Check**
+
+The following fields make up the details provided for each check in the array of checks within the JSON body.
+
+Field          | Type     | Required | Description
+---------------|----------|----------|--------------
+`name`         | `string` | Yes      | Name of the check
+`status`       | `string` | Yes      | The [status of the check](#check-statuses)
+`status_code`  | `int`    | No       | The status code returned by the external service (only for use with external http checks)
+`message`      | `string` | Yes      | Brief description of the status (i.e. `OK` or `received 500 from 'dp-dataset-api'`)
+`last_checked` | `ISO8601`<sup>1</sup> | Yes | The last time the check was run
+`last_success` | `ISO8601`<sup>1</sup> | Yes | The time of the last successful check
```

Comment:
While these are required, is it worth identifying what a default value would be during startup, when there's never been a success or failure?